src/main/java/dev/openfeature/sdk/ProviderRepository.java

```diff
@@ -81,39 +106,72 @@ public void setProvider(String clientName,
+            Consumer<FeatureProvider> afterSet,
+            Consumer<FeatureProvider> afterInit,
+            Consumer<FeatureProvider> afterShutdown,
+            BiConsumer<FeatureProvider, String> afterError) {
+        setProvider(clientName, provider, afterSet, afterInit, afterShutdown, afterError, true);
     }
 
     private void initializeProvider(@Nullable String clientName,
             FeatureProvider newProvider,
             Consumer<FeatureProvider> afterSet,
             Consumer<FeatureProvider> afterInit,
             Consumer<FeatureProvider> afterShutdown,
-            BiConsumer<FeatureProvider, String> afterError) {
+            BiConsumer<FeatureProvider, String> afterError,
+            boolean waitForInit) {
         // provider is set immediately, on this thread
         FeatureProvider oldProvider = clientName != null
-                ? this.providers.put(clientName, newProvider)
-                : this.defaultProvider.getAndSet(newProvider);
+            ? this.providers.put(clientName, newProvider)
+            : this.defaultProvider.getAndSet(newProvider);
         afterSet.accept(newProvider);
-        taskExecutor.submit(() -> {
-            // initialization happens in a different thread
-            try {
-                if (ProviderState.NOT_READY.equals(newProvider.getState())) {
-                    newProvider.initialize(OpenFeatureAPI.getInstance().getEvaluationContext());
-                    afterInit.accept(newProvider);
-                }
-                shutDownOld(oldProvider, afterShutdown);
-            } catch (Exception e) {
-                log.error("Exception when initializing feature provider {}", newProvider.getClass().getName(), e);
-                afterError.accept(newProvider, e.getMessage());
+        if (waitForInit) {
+            initializeProvider(newProvider, afterInit, afterShutdown, afterError, oldProvider);
+        } else {
+            taskExecutor.submit(() -> {
+                // initialization happens in a different thread
+                initializeProvider(newProvider, afterInit, afterShutdown, afterError, oldProvider);
+            });
+        }
+    }
+
+    private void initializeProvider(FeatureProvider newProvider,
+            Consumer<FeatureProvider> afterInit,
+            Consumer<FeatureProvider> afterShutdown,
+            BiConsumer<FeatureProvider, String> afterError,
+            FeatureProvider oldProvider) {
+        try {
+            if (ProviderState.NOT_READY.equals(newProvider.getState())) {
+                newProvider.initialize(OpenFeatureAPI.getInstance().getEvaluationContext());
+                afterInit.accept(newProvider);
             }
-        });
+            shutDownOld(oldProvider, afterShutdown);
+        } catch (Exception e) {
+            log.error("Exception when initializing feature provider {}", newProvider.getClass().getName(), e);
+            afterError.accept(newProvider, e.getMessage());
+        }
```

Comment:
Could you either combine both `initializeProvider` methods, or name them differently? I find overloading these a bit confusing in this case since they are doing different things (one is actually initializing, while the other is just deciding to block or not).